### PR TITLE
[BEAM-4008] Futurize utils subpackage

### PR DESCRIPTION
**Pull request by RobbeSneyders**
This pull request prepares the utils subpackage for Python 3 support. This PR is part of a series in which all subpackages will be updated using the same approach.
This approach has been documented [here](https://docs.google.com/document/d/1xDG0MWVlDKDPu_IW9gtMvxi2S9I0GB0VDTkPhjXT0nE) and the first pull request in the series (Futurize coders subpackage) demonstrating this approach can be found at #5053.

R: @aaltay @tvalentyn

### CODE REVIEW

File: sdks/python/apache_beam/utils/windowed_value.py [hunk 1]
```diff
@@ -178,33 +182,14 @@ def __repr__(self):
...
-    if type(left) is not type(right):
-      return cmp(type(left), type(right))
+  def _key(self):
+    return self.value, self.timestamp_micros, self.windows, self.pane_info
 
-    # TODO(robertwb): Avoid the type checks?
-    # Returns False (0) if equal, and True (1) if not.
-    return not WindowedValue._typed_eq(left, right)
+  def __eq__(self, other):
+    return type(self) == type(other) and self._key() == other._key()
```
Comment (RobbeSneyders) [inline sdks/python/apache_beam/utils/windowed_value.py]: @robertwb should \_\_eq\_\_ be defined in the .pxd file or is this optimized automatically?

Comment (robertwb) [reply sdks/python/apache_beam/utils/windowed_value.py]: __eq__ is a builtin (slots) method that is optimized automatically.

File: sdks/python/apache_beam/utils/timestamp.py [hunk 2]
```diff
@@ -143,11 +148,35 @@ def __int__(self):
     # Note that the returned value may have lost precision.
     return self.micros // 1000000
 
-  def __cmp__(self, other):
-    # Allow comparisons between Duration and Timestamp values.
```
Comment (tvalentyn) [inline sdks/python/apache_beam/utils/timestamp.py]: Let's keep this comment to explain the reason for the casting.

File: sdks/python/apache_beam/utils/timestamp.py [hunk 3]
```diff
@@ -143,11 +148,35 @@ def __int__(self):
...
+  def __eq__(self, other):
```
Comment (tvalentyn) [inline sdks/python/apache_beam/utils/timestamp.py]: How about we keep `__eq__` and `__lt__`, remove the rest and decorate the class with `@functools.total_ordering`?

File: sdks/python/apache_beam/utils/windowed_value.py [hunk 4]
```diff
@@ -178,33 +182,14 @@ def __repr__(self):
...
 
-  @staticmethod
-  def _typed_eq(left, right):
-    return (left.timestamp_micros == right.timestamp_micros
-            and left.value == right.value
-            and left.windows == right.windows
-            and left.pane_info == right.pane_info)
+  def __hash__(self):
+    return hash((type(self),) + self._key())
```
Comment (tvalentyn) [inline sdks/python/apache_beam/utils/windowed_value.py]: I think hash(self._key()) would be sufficient.

File: sdks/python/apache_beam/utils/profiler.py [hunk 5]
```diff
@@ -20,21 +20,19 @@
 For internal use only; no backwards-compatibility guarantees.
 """
 
-import cProfile
+from __future__ import absolute_import
+
+import cProfile  # pylint: disable=bad-python3-import
```
Comment (tvalentyn) [inline sdks/python/apache_beam/utils/profiler.py]: For my education, why does py3 lint complain here?

File: sdks/python/apache_beam/utils/windowed_value.py [hunk 6]
```diff
@@ -178,33 +182,14 @@ def __repr__(self):
...
+  def __eq__(self, other):
+    return type(self) == type(other) and self._key() == other._key()
 
-  @staticmethod
-  def _typed_eq(left, right):
-    return (left.timestamp_micros == right.timestamp_micros
-            and left.value == right.value
-            and left.windows == right.windows
-            and left.pane_info == right.pane_info)
+  def __hash__(self):
```
Comment (tvalentyn) [inline sdks/python/apache_beam/utils/windowed_value.py]: Thanks for your patience for awaiting the feedback, @RobbeSneyders.

I doublechecked  performance of __hash__  using a microbenchmark. Previous implementation of __hash__ appears to be at least 7.5% faster, possibly due to overheads to create a tuple, and/or the way Cython compiles it.

I suggest we revert to previous implementation of __hash__. For consistency  we can also implement __eq__  by comparing fields directly.  

My microbenchmark can be found in https://github.com/apache/beam/compare/master...tvalentyn:utils_futurization_benchmark, see https://github.com/tvalentyn/beam/commit/67d3df0f1248f43feb84503361f0071efc560fc2. It depends on the mini-framework that's being finalized in https://github.com/apache/beam/pull/5565.

CC: @robertwb

Comment (RobbeSneyders) [reply sdks/python/apache_beam/utils/windowed_value.py]: Thanks for adding the benchmarking, this is very useful!
I just ran the benchmark myself and got a performance regression of 6% for reading and 15% for adding to a dict.

I will revert the changes.

File: sdks/python/apache_beam/utils/windowed_value.py [hunk 7]
```diff
@@ -178,34 +182,22 @@ def __repr__(self):
         self.windows,
         self.pane_info)
 
+  def _key(self):
```
Comment (tvalentyn) [inline sdks/python/apache_beam/utils/windowed_value.py]: We don't need this anymore.

File: sdks/python/apache_beam/utils/windowed_value.py [hunk 8]
```diff
@@ -178,34 +182,22 @@ def __repr__(self):
...
+    return self.value, self.timestamp_micros, self.windows, self.pane_info
+
+  def __eq__(self, other):
+    return (type(self) == type(other)
+            and self.timestamp_micros == other.timestamp_micros
+            and self.value == self.value
```
Comment (tvalentyn) [inline sdks/python/apache_beam/utils/windowed_value.py]: s/self/other in right hand side of the equality.

Comment (RobbeSneyders) [reply sdks/python/apache_beam/utils/windowed_value.py]: Sorry, took this straight from the benchmark and didn't notice it.

Comment (tvalentyn) [reply sdks/python/apache_beam/utils/windowed_value.py]: No worries, thank you.

**Comment by tvalentyn** [comment]
LGTM, thank you, @RobbeSneyders!

**Pull request merged**